Tests: provide a way to generate a simple metadata set

MVrachev opened the pull request:
Related to: #1806

**Description of the changes being introduced by the pull request**:

I created a new script called "generate_md.py" which can be used
to easily generate a repository. Additionally, I created a new
test file making sure that the locally stored metadata files and
the newly generated metadata roles are the same.
This will allow us to test that we are not changing the metadata
file structure when making changes.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Provide a way to generate a simple repository

I created a new script called "generate_md.py" which can be used
to easily generate a repository. Additionally, I created a new
test file making sure that the locally stored metadata files and
the newly generated metadata roles are the same.
This will allow us to test that we are not changing the metadata
file structure when making changes.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I am not sure how to resolve the mypy issue: https://github.com/theupdateframework/python-tuf/runs/5467473728?check_suite_focus=true
@jku any suggestions?

jku commented:
> I am not sure how to resolve the mypy issue: https://github.com/theupdateframework/python-tuf/runs/5467473728?check_suite_focus=true @jku any suggestions?

Directories do not become importable packages automatically. You need at least an empty `__init__.py`

MVrachev commented:
@jku I fixed the mypy issues.
Additionally, I had to add the second commit as on the previous CI run the test failed on all Windows venvs https://github.com/theupdateframework/python-tuf/runs/5467473050?check_suite_focus=true).
The second commit aims to achieve consistency on all operating systems.

The downside to this is that the metadata files won't be human-readable.

jku commented:
> The second commit aims to achieve consistency on all operating systems.
> 
> The downside to this is that the metadata files won't be human-readable.

That sounds highly undesirable... not to mention that it sounds like a bug somewhere. What is the inconsistency?

jku reviewed (changes requested):
Hmm, turns out I have opinions about this. I think we can both
* reduce complexity and
* make this more prepared for future additions


Some suggestions:
* keys that are meant for use in multiple places should not be named for their role (this is the reason I don't want to use the legacy test keys). they should be named for the type of key they are (like "ed25519_1") so that individual tests, or in this case generating functions, can decide what they use the key for.
* The CLI complexity is unused: let's not have arguments unless we really need them. Currently neither argument is needed
* The generated data (files/roles) are now hard coded in four different functions: if I next want to add a single file, I would have to modify all four functions. This cannot be the optimal design. I think the object does not need attributes at all: `generate_repository()` could just do everything it needs to (either write the files it wants to write or validate the same files, depending on the argument): once the function is finished, no-one else needs to know which files it touched.
* generate_repository() calls the data it generates a repository and I get the aim of providing examples of python-tuf repositories... but I think it may be premature:
  * Most importantly **we don't actually check that it's a valid repository**. It looks like one but I can only tell because it's the simplest possible, anything more complex is going to be hard to estimate. The filenames are at least incorrect for a consistent_snapshot repository
  * what if I want to add e.g. a file that tests all keytypes we support? Do I have to create a complete repository to add that one file? Looking at the metadata types, timestamp probably only needs one test case but targets might need multiple different files...
  * Would a better approach be to have one or more methods to generate just a bunch of independent metadata files that are not called a repository, and later we add repository support when we can reasonably test it?

Commit:
gitattributes: make all JSON files end with LF

A really specific bug occurred on CI runs on all Windows machines
https://github.com/theupdateframework/python-tuf/runs/5467473050?check_suite_focus=true
where we weren't able to verify that what was generated is the same
as the stored on Git.

After research with Jussi, we found out that the problem comes not
from the content of the file that was generated, but because on Windows
Git proactively replaced all line endings for text files with CRLF symbol
("\r") this made the locally stored JSON files different from the one
generated.

We want to make sure such bugs doesn't occur again and that's why we
disable this behavior for all JSON files.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku I think I have simplified the test script a lot.
The test doesn't require context and it should be a lot better.

MVrachev commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+        md_targets: Metadata[Targets] = Metadata(targets, {})
+
+        for md in [md_root, md_timestamp, md_snapshot, md_targets]:
+            assert isinstance(md, Metadata)
```
This check is required from mypy otherwise there is a warning:
`
tests/generated_data/generate_md.py:142: error: Argument 1 to "verify_generation" of "MetadataGenerator" has incompatible type "object"; expected "Metadata[Any]"  [arg-type]
`

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+)
+from tuf.api.serialization.json import JSONSerializer
+
+# Hardcode keys and expiry time to achieve reproducibility.
```
Maybe hardcode expiry here to match the comment?

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+            "private": "7e2e751145d1b22f6e40d4ba2aa47158207acfd3c003f1cbd5a08141dfc22a15",
+        },
+    },
+}
```
You could condense this a bit, but having a list of (keyid, public val, private val)-tuples and construct the dicts below in a loop. You could even omit keyid in the tuple and compute it in the loop.

MVrachev replied:
I followed your suggestions and made three lists `public_values`, `private_values` and `keyids`.
I don't want to calculate the `keyids` as that is a possible place where a change inside `securesystemslib` will result in our tests failing which I am not sure we want.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+    "ed25519_2": Key.from_securesystemslib_key(keys_dict["ed25519_2"]),
+    "ed25519_3": Key.from_securesystemslib_key(keys_dict["ed25519_3"]),
+    "ed25519_4": Key.from_securesystemslib_key(keys_dict["ed25519_4"]),
+}
```
Looks like a use case for a loop / dict comprehension.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+}
+
+
+class MetadataGenerator:
```
Why use a class? Will there be multiple `MetadataGenerator` objects alongside each other? Wouldn't it be enough to make `generate_all_files` a normal function that takes all constant values from module globals (it already does for keys) and all variable values from function arguments?

MVrachev replied:
I used class mostly because I can reuse variables.
I just wanted to escape from the idea of using script globals that are not hardcoded and generally for me it's easier to follow knowing which variables are reused across functions when you have the `self` prefix.
We have discussed with @jku that it's most likely we will enhance this script generating more variations of the same roles and that could be another reason why a class could be better.

If you still prefer to remove the class I will.

lukpueh replied:
> If you still prefer to remove the class I will.

I do because I can't see an advantage of using a class. But I leave it up to you.

MVrachev replied:
I decided to remove the class.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+        """Generate a new repository and optionally verify it.
+
+        Args:
+            dump: Wheter to dump the newly generated files inside self.dir.
```
Is there a use case where the files are not dumped?

MVrachev replied:
Yes, when we are testing that functionality https://github.com/MVrachev/tuf/blob/45f20beaa83bd2b8b6dbd44164e1a1b7883bd741/tests/test_metadata_generation.py#L20.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,153 @@
+        # Dump metadata files in a human readable format.
+        self.serializer = JSONSerializer()
+
+    def verify_generation(self, md: Metadata, path: str) -> None:
```
🎉  I like this function. It's a good example for factoring out functionality that could be in `generate_all_files`, so that it can also be used re-used by other metadata generating functions as well.

lukpueh reviewed (commented):
I commented a couple of observations/questions, mostly about further simplifying the script.
But maybe I'm missing the big picture. In what direction should the script scale? 

I'd probably make it just a collection of `generate_*_metadata` functions, plus helper functions to be re-used by those generation function, like  your verification function.

MVrachev commented:
> I commented a couple of observations/questions, mostly about further simplifying the script. But maybe I'm missing the big picture. In what direction should the script scale?
> 
> I'd probably make it just a collection of `generate_*_metadata` functions, plus helper functions to be re-used by those generation function, like your verification function.

I did it that way before and @jku mentioned that there is no sense in having separate functions now when we generate only one variation per role.
If we enhance this script to generate more variations (as we plan to) it will make sense to have separate `generate_*_metadata` functions.
@jku has mentioned something like this in his last points here https://github.com/theupdateframework/python-tuf/issues/1806#issuecomment-1060417793.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,146 @@
+]
+
+keys: Dict[str, Key] = {}
+for i in range(0, 4):
```
Hint: `range(0, 4) == range(4)`

lukpueh commented:
> > I'd probably make it just a collection of `generate_*_metadata` functions, plus helper functions to be re-used by those generation function, like your verification function.
> 
> I did it that way before and @jku mentioned that there is no sense in having separate functions now when we generate only one variation per role.

Oh, I probably wasn't clear. I didn't mean that you should split `generate_all_files` now. I was talking about the future when you have multiple functions that each generate a set of metadata files for a specific use case. So I think we are on the same page here.

MVrachev commented:
@lukpueh I updated the pr with removing the class and a couple of small simplifications.
@jku and @lukpueh it's ready for another review.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,137 @@
+    os.mkdir(OUT_DIR)
+
+SERIALIZER = JSONSerializer()
+key_info = copy.deepcopy(keys["ed25519_0"].to_securesystemslib_key())
```
Why deepcopy?

MVrachev replied:
Because at `keys["ed25519_0"]` we have the dictionary information of a dict containing a dictionary of a dictionary.
More precisely the keyval of a key is again a dictionary and a simple `copy.copy()` doesn't work here.

lukpueh replied:
What do you mean by "doesn't work here"? And why do you need to copy it at all?

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,137 @@
+
+    for md in [md_root, md_timestamp, md_snapshot, md_targets]:
+        assert isinstance(md, Metadata)
+        md.sign(SIGNER)
```
Why do you sign all metadata files with the root key? Why not sign timestamp with timestamp, snapshot with snapshot and targets with targets keys?

MVrachev replied:
Changed that.
Initially (while there was a class), it was to reuse the signer.

lukpueh replied:
I am not sure I understand. So initially you wanted to sign with one key on purpose?

MVrachev replied:
Yea, because the files are not supposed to be read as a stand-alone repository, but as separate independent files, and as such we can reuse the same key as many times as we want.
The only reason I defined 4 keys is because of the root roles.

lukpueh reviewed (commented):
Cool! Thanks for the updates, Martin! I think it's easier to review and use now, without making it much harder to scale later. Here are two more comments about signing...

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,140 @@
+    targets = Targets(1, SPEC_VERSION, EXPIRY, {})
+    md_targets: Metadata[Targets] = Metadata(targets, {})
+
+    i = 0
```
hint: `enumerate()` gives you a running index.

MVrachev replied:
I remembered it just when I sent it, thanks.

lukpueh commented on tests/generated_data/generate_md.py:
```diff
@@ -0,0 +1,140 @@
+    for md in [md_root, md_timestamp, md_snapshot, md_targets]:
+        assert isinstance(md, Metadata)
+        key_info = copy.deepcopy((keys[f"ed25519_{i}"].to_securesystemslib_key()))
+        key_info["keyval"]["private"] = private_values[0]
```
```suggestion
        key_info["keyval"]["private"] = private_values[i]
```
?

lukpueh replied:
I think you could avoid `deepcopy`, `to_securesystemslib_key` and the assignment of `private_values[i]` if you use the private key dictionary defined above (before passing it to `from_securesystemslib_key`).

MVrachev commented:
Not sure why, but GitHub doesn't see my update which exists on the same branch on my fork...
Will close and reopen the pr.

MVrachev commented:
@lukpueh because of a GitHub bug the new changes inside test-static-data-generation branch of my fork
weren't noticed. I closed it, but there was no option to reopen it.
That's why I created #1915.
